// firebase-function/lib/src/testUtils/FirebaseApp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseApp = void 0;
const app_1 = require("firebase/app");
const database_1 = require("firebase/database");
const auth_1 = require("firebase/auth");
const FirebaseFunctions_1 = require("./FirebaseFunctions");
const FirebaseDatabase_1 = require("./FirebaseDatabase");
const FirebaseAuth_1 = require("./FirebaseAuth");
const functions_1 = require("firebase/functions");
class FirebaseApp {
    cryptionKeys;
    callSecretKey;
    _functions;
    _database;
    _auth;
    constructor(options, cryptionKeys, callSecretKey, config) {
        this.cryptionKeys = cryptionKeys;
        this.callSecretKey = callSecretKey;
        const app = (0, app_1.initializeApp)(options, config?.name);
        this._functions = (0, functions_1.getFunctions)(app, config?.functionsRegion);
        this._database = (0, database_1.getDatabase)(app, config?.databaseUrl);
        this._auth = (0, auth_1.getAuth)(app);
    }
    get functions() {
        return new FirebaseFunctions_1.FirebaseFunctions(this._functions, this.cryptionKeys, this.callSecretKey);
    }
    get database() {
        return new FirebaseDatabase_1.FirebaseDatabase(this._database, this.cryptionKeys);
    }
    get auth() {
        return new FirebaseAuth_1.FirebaseAuth(this._auth);
    }
}
exports.FirebaseApp = FirebaseApp;
